### PR TITLE
[No QA][TS migration] Migrate 'Visibility' lib to TypeScript

### PR DESCRIPTION
**Pull request by Skalakid**
### Fixed Issues

$ https://github.com/Expensify/App/issues/24844
PROPOSAL: https://github.com/Expensify/App/issues/24844

### Tests

1. Log in
2. Enter any chat room
3. Open console and select `verbose` logs
4. Leave application and verify if "Flushing logs as app is going inactive - {}" appears
5. Come back to the app and verify if many new logs appear

- [x] Verify that no errors appear in the JS console

### Offline tests

Same as Tests
### QA Steps

Same as tests
- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://github.com/Expensify/App/assets/39538890/bdb620c9-aad1-4e22-b004-6573704ee520

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://github.com/Expensify/App/assets/39538890/b544879d-ff2f-461e-933b-96c0dc567f5a

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://github.com/Expensify/App/assets/39538890/f609b17c-a2b3-42ed-8270-67a880a3f41d

</details>

<details>
<summary>Desktop</summary>

https://github.com/Expensify/App/assets/39538890/99975429-75e6-4cf6-8a90-6a553965410d

</details>

<details>
<summary>iOS</summary>

https://github.com/Expensify/App/assets/39538890/433af8c1-e14f-4328-92a4-bbfea12bf596

</details>

<details>
<summary>Android</summary>

https://github.com/Expensify/App/assets/39538890/80642122-9658-4d2a-aa44-7b816cfb127b

</details>

### CODE REVIEW

File: src/libs/Visibility/index.desktop.ts [hunk 1]
```diff
@@ -4,28 +4,20 @@ import ELECTRON_EVENTS from '../../../desktop/ELECTRON_EVENTS';
  * Detects whether the app is visible or not. Electron supports document.visibilityState,
  * but switching to another app while Electron is partially occluded will not trigger a state of hidden
  * so we ask the main process synchronously whether the BrowserWindow.isFocused()
- *
- * @returns {Boolean}
  */
 function isVisible() {
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-return
```
Comment (blazejkustra) [inline src/libs/Visibility/index.desktop.ts:9]: I think electron is typed as 'any', we have to change this @Skalakid

File: src/libs/Visibility/index.desktop.ts [hunk 2]
```diff
@@ -4,28 +4,20 @@ import ELECTRON_EVENTS from '../../../desktop/ELECTRON_EVENTS';
...
     return window.electron.sendSync(ELECTRON_EVENTS.REQUEST_VISIBILITY);
 }
 
-/**
- * @returns {Boolean}
- */
 function hasFocus() {
```
Comment (blazejkustra) [inline src/libs/Visibility/index.desktop.ts:13]: Please create `types.ts` file with shared types, similar to this [PR](https://github.com/Expensify/App/pull/27022/files)

File: src/types/modules/electron.d.ts [hunk 3]
```diff
@@ -0,0 +1,9 @@
+declare global {
+    // eslint-disable-next-line @typescript-eslint/consistent-type-definitions
+    interface Window {
+        electron: Electron;
```
Comment (blazejkustra) [inline src/types/modules/electron.d.ts:4]: In my IDE Electron is of type any

**Review by blazejkustra — CHANGES_REQUESTED** [review]

File: src/types/modules/electron.d.ts [hunk 4]
```diff
@@ -0,0 +1,9 @@
+declare global {
+    // eslint-disable-next-line @typescript-eslint/consistent-type-definitions
+    interface Window {
+        electron: Electron.IpcRenderer;
```
Comment (blazejkustra) [inline src/types/modules/electron.d.ts:4]: NAB: This type might be too extensive, we should probably create a type based on the implementation [here](https://github.com/software-mansion-labs/expensify-app-fork/blob/fd4fa9f6d4a4ecaeb437589d153495782e29611c/desktop/contextBridge.js). [Article](https://www.jsgarden.co/blog/how-to-handle-electron-ipc-events-with-typescript) that explains it well.

**Review by blazejkustra — APPROVED** [review]
Either way LGTM.

File: src/libs/Visibility/index.desktop.ts [hunk 5]
```diff
@@ -1,31 +1,19 @@
...
  * Detects whether the app is visible or not. Electron supports document.visibilityState,
  * but switching to another app while Electron is partially occluded will not trigger a state of hidden
  * so we ask the main process synchronously whether the BrowserWindow.isFocused()
- *
- * @returns {Boolean}
  */
-function isVisible() {
-    return window.electron.sendSync(ELECTRON_EVENTS.REQUEST_VISIBILITY);
-}
+const isVisible: IsVisible = () => !!window.electron.sendSync(ELECTRON_EVENTS.REQUEST_VISIBILITY);
```
Comment (s77rt) [inline src/libs/Visibility/index.desktop.ts:9]: I don't think we had to remove the named function here.

`function isVisible(): boolean` would be fine I guess.

**Review by s77rt — APPROVED** [review]
Approving for engineer review

File: src/libs/Visibility/index.desktop.ts (hunk 5 above)
Comment (blazejkustra) [reply src/libs/Visibility/index.desktop.ts:9]: From TS guidelines:

> In modules with platform-specific implementations, create `types.ts` to define shared types. For platform-specific implementations, always define shared types that complies with all variants.

That's why `IsVisible` type was created in the first place and used here. I think the change was necessary.

**Review by fabioh8010 — APPROVED** [review]
LGTM, CK still need to do the checklist.

**Comment by fabioh8010** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos

<details>
<summary>Web</summary>


https://github.com/Expensify/App/assets/20051562/ee3da277-ef37-4ade-be62-447e84fd14f9


https://github.com/Expensify/App/assets/20051562/f932e39c-313f-49f3-9e0b-a1c698f53894



</details>

<details>
<summary>Desktop</summary>


https://github.com/Expensify/App/assets/20051562/af2a32f9-6040-4770-bac7-f1f58c50a782



</details>

<details>
<summary>Mobile Web - Chrome</summary>


https://github.com/Expensify/App/assets/20051562/7c7d2dc5-3f03-4e05-9bb0-0a5f30eb9eb6



</details>

<details>
<summary>Android</summary>


https://github.com/Expensify/App/assets/20051562/b976b768-69ec-4ab0-a2f9-9f5181465271



</details>

<details>
<summary>Mobile Web - Safari</summary>


https://github.com/Expensify/App/assets/20051562/fb210eca-518f-458a-8fcd-6c22e78045c1



</details>

<details>
<summary>iOS</summary>


https://github.com/Expensify/App/assets/20051562/71dff2bd-673d-4cd7-88a1-299ab7eb8fd5



</details>

**Review by fabioh8010 — APPROVED** [review]

File: src/types/modules/electron.d.ts [hunk 6]
```diff
@@ -0,0 +1,18 @@
+// TODO: Move this type to desktop/contextBridge.js once it is converted to TS
```
[the remark this answers is not in the document]
Comment (Skalakid) [reply src/types/modules/electron.d.ts:1]: Yes, here is a link to desktop files migration issue: https://github.com/Expensify/App/issues/25333

[the remark this answers is not in the document]
Comment (flodnv) [reply src/types/modules/electron.d.ts:1]: Cool thanks, we could also link it here for clarity

**Review by flodnv — APPROVED** [review]

**Pull request merged**